src/middleware.ts

```diff
@@ -259,43 +296,60 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
-
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+  // The state that was loaded from the storage and is set via set() within this
+  // middleware would later be overridden with the initial state by create(). To
+  // avoid this we merge the state from localStorage into the initial state. This
+  // way the state gets overridden with the state loaded from storage.
+  // TODO: find a better solution for this
+  let stateFromStorage
+  ;(() => {
+    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
+    // bind is used to avoid `TypeError: Illegal invocation` error
+    let storageValuePending = toThenable(storage.getItem.bind(storage))(name)
+    storageValuePending
+      .then((storageValue) => {
+        if (storageValue) {
+          return deserialize(storageValue)
+        }
+      })
+      .then((deserializedStorageValue) => {
+        if (deserializedStorageValue) {
+          if (deserializedStorageValue.version !== version) {
+            if (!migrate) return
+            stateFromStorage = migrate(
+              deserializedStorageValue.state,
+              deserializedStorageValue.version
+            )
+            return stateFromStorage
```

Comment:
```suggestion
            if (migrate) {
              stateFromStorage = migrate(
                deserializedStorageValue.state,
                deserializedStorageValue.version
              )
              return stateFromStorage
            }
```
Seems more consistent with other code.